include/blackhole/detail/sink/asynchronous.hpp

```diff
@@ -0,0 +1,80 @@
+#include <atomic>
+#include <thread>
+
+#include <cds/container/vyukov_mpmc_cycle_queue.h>
+
+#include "blackhole/sink.hpp"
+
+#include "blackhole/detail/recordbuf.hpp"
+
+namespace blackhole {
+inline namespace v1 {
+namespace experimental {
+namespace sink {
+
+using detail::recordbuf_t;
+
+/// I can imagine: drop, sleep, wait.
+class overflow_policy_t {
+public:
+    enum class action_t {
+        retry,
+        drop
+    };
+
+public:
+    virtual ~overflow_policy_t() {}
+
+    /// Handles record queue overflow.
+    ///
+    /// This method is called when the queue is unable to enqueue more items. It's okay to throw
+    /// exceptions from here, they will be propagated directly to the sink caller.
+    virtual auto overflow() -> action_t = 0;
+
+    virtual auto wakeup() -> void = 0;
+};
+
+class overflow_policy_factory_t {
+public:
+    auto create(const std::string& name) const -> std::unique_ptr<overflow_policy_t>;
+};
+
+class asynchronous_t : public sink_t {
+    struct value_type {
+        recordbuf_t record;
+        std::string message;
+    };
+
+    typedef cds::container::VyukovMPSCCycleQueue<value_type> queue_type;
+
+    queue_type queue;
+    std::atomic<bool> stopped;
+    std::unique_ptr<sink_t> wrapped;
+
+    std::unique_ptr<overflow_policy_t> overflow_policy;
+
+    std::thread thread;
+
+public:
+    asynchronous_t(std::unique_ptr<sink_t> wrapped, std::size_t factor = 10);
+
+    // TODO: Full customization.
+    asynchronous_t(std::unique_ptr<sink_t> sink,
```

Comment:
umm, this ctor includes all parameters from the previous one. Do you really need 2 different ctors?